lib/google/gax/paged_enumerable.rb

```diff
@@ -0,0 +1,200 @@
+    #   while some_condition()
+    #     page = paged_enumerable.page
+    #     do_something(page)
+    #     break if paged_enumerable.next_page?
+    #     paged_enumerable.next_page
+    #   end
+    #
+    class PagedEnumerable
+      include Enumerable
+
+      # @attribute [r] page
+      #   @return [Page] The current page object.
+      attr_reader :page
+
+      # @param request_page_token_field [String]
+      #   The name of the field in request which will have the page token.
+      # @param response_page_token_field [String]
+      #   The name of the field in the response which holds the next page token.
+      # @param resource_field [String]
+      #   The name of the field in the response which holds the resources.
+      def initialize(api_call, request, response, options: nil,
+                     resource_field: nil)
+        @api_call = api_call
+        @request = request
+        @response = response
+        @options = options
+        @resource_field = resource_field || determine_resource_field(response)
+
+        @page = Page.new(@response, @resource_field)
+      end
+
+      # Iterate over the resources.
+      # @yield [Object] Gives the resource objects in the stream.
+      # @raise [RuntimeError] if it's not started yet.
+      def each
+        return enum_for :each unless block_given?
+
+        each_page do |page|
+          page.each do |obj|
+            yield obj
+          end
+        end
+      end
+
+      # Iterate over the pages.
+      # @yield [Page] Gives the pages in the stream.
+      # @raise [GaxError] if it's not started yet.
+      def each_page
+        return enum_for :each_page unless block_given?
+
+        yield @page
+        loop do
+          break unless next_page?
+          yield next_page
+        end
+      end
+
+      # True if it has the next page.
+      def next_page?
+        @page.next_page_token?
+      end
+
+      # Update the response in the current page.
+      # @return [Page] the new page object.
+      def next_page
+        return unless next_page?
+
+        next_request = @request.dup
+        next_request.page_token = @page.next_page_token
+        next_response = @api_call.call(next_request, @options)
+
+        @page = Page.new(next_response, @resource_field)
+      end
+
+      # The page token to be used for the next API call.
+      # @return [String]
+      def next_page_token
+        @page.next_page_token
+      end
+
+      # The current response object, for the current page.
+      # @return [Object]
+      def response
+        @page.response
+      end
+
+      private
+
+      # This inspects a response object and pulls out the first repeated field.
+      def determine_resource_field(response)
+        # Find the first repeated FieldDescriptor on the response Descriptor
+        # Find uses the order of the FieldDescriptor on the Descriptor
+        field = response.class.descriptor.find { |f| f.label == :repeated }
```

Comment:
worth putting a similar sanity check in the initializer so we know the field exists and is of the correct format?

Reply:
Yeah, this is partially why I can see us not passing the `resource_field` at all, and always looking it up.